File: 0322-coin-change/0322-coin-change.java
// Memoization
// class Solution {
//     public int numCoins(int idx, int[] coins, int amt, int[][] dp) {
//         if(amt == 0) return 0;
//         if(idx == 0) {
//             if(amt % coins[idx] == 0) return amt/coins[idx];
//             else return (int) 1e9;
//         }
//         if(dp[idx][amt] != -1) return dp[idx][amt];
//         int notTake = 0 + numCoins(idx-1, coins, amt, dp);
//         int take = (int) 1e9;
//         if(coins[idx] <= amt) {
//             take = 1 + numCoins(idx, coins, amt - coins[idx], dp);
//         }

//         return dp[idx][amt] = Math.min(notTake, take);
//     }
//     public int coinChange(int[] coins, int amount) {
//         int n = coins.length;
//         int[][] dp = new int[n][amount + 1];
//         for(int i=0; i<n; i++) {
//             Arrays.fill(dp[i], -1);
//         }
//         int coinNums = numCoins(n-1, coins, amount, dp);
//         if(coinNums == (int) 1e9) return -1;
//         return coinNums;
//     }
// }

// Tabulation
class Solution {
    public int coinChange(int[] coins, int amount) {
        int n = coins.length;
        int[][] dp = new int[n][amount + 1];
        
        for(int i=0; i<n; i++) {
            dp[i][0] = 0;
        }
        for(int i=0; i<=amount; i++) {
            if(i % coins[0] == 0) dp[0][i] = i / coins[0];
            else dp[0][i] = (int) 1e9;
        }

        for(int i=1; i<n; i++) {
            for(int amt=1; amt<=amount; amt++) {
                int notTake = 0 + dp[i - 1][amt];
                int take = (int) 1e9;
                if(coins[i] <= amt) {
                    take = 1 + dp[i][amt - coins[i]];
                }
                dp[i][amt] = Math.min(take, notTake);
            }
        }
        int ans = dp[n-1][amount];
        if(ans == (int) 1e9) return -1;
        return ans;
    }
}

File: 0322-coin-change/0322-coin-change.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        # if amount == 0:
        #     return 0
        # coins.sort()
        # amt_left = amount
        # coin_count = 0
        # for coin in reversed(coins):
        #     if amt_left >= coin:
        #         temp = amt_left // coin
        #         coin_count += temp
        #         amt_left %= coin
        #     else:
        #         continue
        # if amt_left == 0:
        #     return coin_count
        # else:
        #     return -1

        ## above appeoach is wrong as it only considers some possibilities and ignores the possibility that if we do not take the biggest coind then also we can make the amount. Below approach is the DP approach that is perfect.

        n = len(coins)
        dp = [[0 for _ in range(amount + 1)] for _ in range(n + 1)]

        # base case
        for i in range(n):
            dp[i][0] = 0
        
        for j in range(amount + 1):
            if j % coins[0] == 0:
                dp[0][j] = j // coins[0]
            else:
                dp[0][j] = int(1e9)
        
        # DP for loops
        for i in range(1, n):
            for amt in range(1, amount + 1):
                take = int(1e9)
                if(amt >= coins[i]):
                    take = 1 + dp[i][amt - coins[i]]
                notTake = 0 + dp[i-1][amt]
                dp[i][amt] = min(take, notTake)

        if dp[n-1][amount] != int(1e9):
            return dp[n-1][amount]
        else:
            return -1